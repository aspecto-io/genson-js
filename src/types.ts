export enum ValueType {
    Null = 'null',
    Boolean = 'boolean',
    Integer = 'integer',
    Number = 'number',
    String = 'string',
    Object = 'object',
    Array = 'array',
}

export type Schema = {
    type?: ValueType | ValueType[];
    items?: Schema;
    properties?: Record<string, Schema>;
    required?: string[];
    anyOf?: Array<Schema>;
    description?: string;
};

export type SchemaGenOptions = {
    noRequired: boolean;
};

export type SchemaComparisonOptions = {
    ignoreRequired: boolean;
};
