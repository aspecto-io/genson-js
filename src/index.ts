export * from './types';
export { extendSchema, mergeSchemas, createSchema, createCompoundSchema } from './schema-builder';
export * from './schema-comparator';
