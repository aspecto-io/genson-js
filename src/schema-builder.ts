import { Schema, SchemaGenOptions, ValueType } from './types';


const partition = <T>(arr: T[], criteria: (a: T) => boolean) => arr.reduce((acc, i) => {acc[criteria(i) ? 0 : 1].push(i); return acc}, [[], []]);

function createSchemaFor(value: any, options?: SchemaGenOptions): Schema {
    const nullable = options?.restrictiveNulls ? {nullable: false} : {};
    switch (typeof value) {
        case 'number':
            if (Number.isInteger(value)) {
                return { type: ValueType.Integer, ...nullable };
            }
            return { type: ValueType.Number, ...nullable };
        case 'boolean':
            return { type: ValueType.Boolean, ...nullable };
        case 'string':
            return { type: ValueType.String, ...nullable };
        case 'object':
            if (value === null) {
                return { type: ValueType.Null, ...(options?.restrictiveNulls ? { nullable: true } : {}) };
            }
            if (Array.isArray(value)) {
                return createSchemaForArray(value, options);
            }
            return createSchemaForObject(value, options);
    }
}

function createSchemaForArray(arr: Array<any>, options?: SchemaGenOptions): Schema {
    const nullable = options?.restrictiveNulls ? {nullable: false} : {};
    if (arr.length === 0) {
        return { type: ValueType.Array, ...nullable};
    }
    const elementSchemas = arr.map((value) => createSchemaFor(value, options));
    const items = combineSchemas(elementSchemas);
    let minItems;
    if(options?.restrictiveArrays) {
        minItems = options.minItemsOverride ?? arr.length;
    }
    return { type: ValueType.Array, items, ...(minItems && { minItems }), ...nullable };
}

function createSchemaForObject(obj: Object, options?: SchemaGenOptions): Schema {
    const nullable = options?.restrictiveNulls ? {nullable: false} : {};
    const additionalProps = options?.additionalProperties ? {additionalProperties: true} : {};
    const keys = Object.keys(obj);
    if (keys.length === 0) {
        return {
            type: ValueType.Object,
            ...nullable,
            ...additionalProps,
        };
    }
    const properties = Object.entries(obj).reduce((props, [key, val]) => {
        props[key] = createSchemaFor(val, options);
        return props;
    }, {});

    const schema: Schema = { type: ValueType.Object, properties, ...nullable, ...additionalProps };
    if (!options?.noRequired) {
        schema.required = keys;
    }
    return schema;
}

function combineSchemas(schemas: Schema[], options?: SchemaGenOptions): Schema {
    const schemasByType: Record<ValueType, Schema[]> = {
        [ValueType.Null]: [],
        [ValueType.Boolean]: [],
        [ValueType.Integer]: [],
        [ValueType.Number]: [],
        [ValueType.String]: [],
        [ValueType.Array]: [],
        [ValueType.Object]: [],
    };

    const unwrappedSchemas = unwrapSchemas(schemas);
    for (const unwrappedSchema of unwrappedSchemas) {
        const type = unwrappedSchema.type as ValueType;
        if (schemasByType[type].length === 0 || isContainerSchema(unwrappedSchema)) {
            schemasByType[type].push(unwrappedSchema);
        }
    }

    const resultSchemasByType: Record<ValueType, Schema> = {
        [ValueType.Null]: schemasByType[ValueType.Null][0],
        [ValueType.Boolean]: schemasByType[ValueType.Boolean][0],
        [ValueType.Number]: schemasByType[ValueType.Number][0],
        [ValueType.Integer]: schemasByType[ValueType.Integer][0],
        [ValueType.String]: schemasByType[ValueType.String][0],
        [ValueType.Array]: combineArraySchemas(schemasByType[ValueType.Array], options),
        [ValueType.Object]: combineObjectSchemas(schemasByType[ValueType.Object], options),
    };

    if (resultSchemasByType[ValueType.Number]) {
        // if at least one value is float, others can be floats too
        delete resultSchemasByType[ValueType.Integer];
    }

    const schemasFound = Object.values(resultSchemasByType).filter(Boolean);
    const multiType = schemasFound.length > 1;
    if (multiType) {
        return wrapAnyOfSchema({ anyOf: schemasFound }, options);
    }
    return schemasFound[0] as Schema;
}

function combineArraySchemas(schemas: Schema[], options?: SchemaGenOptions): Schema {
    if (!schemas || schemas.length === 0) {
        return undefined;
    }
    const nullableArr: boolean[] = [];
    const itemSchemas: Schema[] = [];
    for (const schema of schemas) {
        nullableArr.push(schema.nullable ?? true);
        if (!schema.items) continue;
        const unwrappedSchemas = unwrapSchema(schema.items);
        itemSchemas.push(...unwrappedSchemas);
    }

    if(options?.restrictiveNulls && !options?.mergeToLessRestrictive) {
        const firstSchema = schemas[0].items;
        const allSame = schemas.reduce((acc, schema) => acc && isSameSchema(firstSchema, schema.items), true);

        if(allSame && nullableArr.indexOf(false) > -1) {
            return {...schemas[0], nullable: false};
        }

        const [nullable, nonNullable] = partition(schemas, s => s.nullable);
        if(nonNullable.length > 0 && nullable.length > 0) {
            return {
                anyOf: [
                    combineArraySchemas(nonNullable, options),
                    combineArraySchemas(nullable, options),
                ]
            }
        }
    }

    if (options?.mergeToLessRestrictive && schemas.length !== schemas.filter(schema => !!schema.items).length) {
        return {
            type: ValueType.Array,
        };
    }

    let minItems = options?.minItemsOverride;
    for (const schema of schemas) {
        if (!schema.minItems) {
            minItems = undefined;
            break;
        }
        if(minItems === undefined) {
            minItems = schema.minItems;
        } else {
            minItems = Math.min(minItems, schema.minItems);
        }
    }

    if (itemSchemas.length === 0) {
        return {
            type: ValueType.Array,
        };
    }

    const items = combineSchemas(itemSchemas);
    return {
        type: ValueType.Array,
        items,
        ...(minItems && (minItems !== 0 ? { minItems } : {})),
        ...(nullableArr.indexOf(false) > -1 ? {nullable: (!!options?.mergeToLessRestrictive && nullableArr.indexOf(true) > -1)} : {}),
    };
}

const isSameSchema = (schema1: Schema, schema2: Schema) => {
    if(schema1 === undefined || schema2 === undefined) {
        return schema1 === schema2;

    }
    if(schema1.type !== schema2.type) {
        return false;
    }
    if(schema1.type === ValueType.Array) {
        return isSameSchema(schema1.items, schema2.items);
    }
    if(schema1.type === ValueType.Object) {
        const keys1 = Object.keys(schema1.properties);
        const keys2 = Object.keys(schema2.properties);
        if(keys1.length !== keys2.length) {
            return false;
        }
        for(const key of keys1) {
            if(!isSameSchema(schema1.properties[key], schema2.properties[key])) {
                return false;
            }
        }
        return true;
    }
    return true;
}

function combineObjectSchemas(schemas: Schema[], options?: SchemaGenOptions): Schema {
    if (!schemas || schemas.length === 0) {
        return undefined;
    }
    const nullableArr: boolean[] = [];
    schemas.forEach((schema) => {
        nullableArr.push(schema.nullable ?? true);
    })
    const additionalProps: boolean[] = [];
    schemas.forEach((schema) => {
        additionalProps.push(schema.additionalProperties ?? true);
    })
    if(options?.restrictiveNulls && !options?.mergeToLessRestrictive) {
        const [nullable, nonNullable] = partition(schemas, s => s.nullable ?? true);
        if(nonNullable.length > 0 && nullable.length > 0) {
            return {
                anyOf: [
                    combineObjectSchemas(nonNullable, options),
                    combineObjectSchemas(nullable, options),
                ]
            }
        }
    }
    const allPropSchemas = schemas.map((s) => s.properties).filter(Boolean);
    if(options?.mergeToLessRestrictive && allPropSchemas.length !== schemas.length) {
        return {
            type: ValueType.Object,
            ...(nullableArr.indexOf(false) > -1 ? {nullable: (!!options?.mergeToLessRestrictive && nullableArr.indexOf(true) > -1)} : {}),
            ...(additionalProps.indexOf(true) > -1 ?  {} : {additionalProperties: false}),
        }
    }
    const schemasByProp: Record<string, Schema[][]> = Object.create(null);
    for (const propSchemas of allPropSchemas) {
        for (const [prop, schema] of Object.entries(propSchemas)) {
            if (!schemasByProp[prop]) {
                schemasByProp[prop] = [];
            }
            const unwrappedSchemas = unwrapSchema(schema);
            schemasByProp[prop].push(unwrappedSchemas);
        }
    }

    if(options?.mergeToLessRestrictive) {
        for (const [prop, subschemas] of Object.entries(schemasByProp)) {
            if(subschemas.length !== schemas.length) {
                delete schemasByProp[prop];
            }
        }
    }

    const properties: Record<string, Schema> = Object.entries(schemasByProp).reduce((props, [prop, schemaArr]) => {
        const schemas = schemaArr.reduce((acc, s) => [...acc, ...s], []);
        if (schemas.length === 1) {
            props[prop] = schemas[0];
        } else {
            const combined = combineSchemas(schemas, options);
            if(!(options?.restrictiveNulls ?? false)) {
                const [nullable, nonNullable] = partition(schemas, s => s.nullable);
                if(nonNullable.length > 0 && nullable.length > 0) {
                    delete combined.nullable;
                }
            }
            props[prop] = combined;
        }
        return props;
    }, {});

    const combinedSchema: Schema = { type: ValueType.Object };

    if (Object.keys(properties).length > 0) {
        combinedSchema.properties = properties;
    }
    if (!options?.noRequired) {
        const required = intersection(schemas.map((s) => s.required || []));
        if (required.length > 0) {
            combinedSchema.required = required;
        }
    }
    if(nullableArr.indexOf(false) > -1) {
        if (!(options?.mergeToLessRestrictive && nullableArr.indexOf(true) > -1)) {
            combinedSchema.nullable = false;
        }
    }

    if(additionalProps.indexOf(true) === -1) {
        combinedSchema.additionalProperties = false;
    }

    return combinedSchema;
}

export function unwrapSchema(schema: Schema): Schema[] {
    if (!schema) return [];
    if (schema.anyOf) {
        return unwrapSchemas(schema.anyOf);
    }
    if (Array.isArray(schema.type)) {
        return schema.type.map((x) => ({ type: x }));
    }
    return [schema];
}

export function unwrapSchemas(schemas: Schema[]): Schema[] {
    if (!schemas || schemas.length === 0) return [];
    return schemas.flatMap((schema) => unwrapSchema(schema));
}

export function wrapAnyOfSchema(schema: Schema, options?: SchemaGenOptions): Schema {
    const simpleSchemas = [];
    const complexSchemas = [];
    let nullable: boolean = undefined;
    for (const subSchema of schema.anyOf) {
        if (isSimpleSchema(subSchema) && !options?.restrictiveNulls) {
            if(Array.isArray(subSchema.type)) {
                simpleSchemas.push(...subSchema.type);
            } else {
                simpleSchemas.push((subSchema as Schema).type);
            }
            nullable = true;
        } else if (isSimpleSchema(subSchema) && options?.restrictiveNulls) {
            if (nullable === undefined || nullable === true) {
                if(Array.isArray(subSchema.type)) {
                    simpleSchemas.push(...subSchema.type);
                } else {
                    simpleSchemas.push((subSchema as Schema).type);
                }
                nullable = true;
            } else {
                complexSchemas.push(subSchema);
            }
        } else if (isSimpleSchemaWithNullable(subSchema) && !options?.restrictiveNulls) {
            if(Array.isArray(subSchema.type)) {
                simpleSchemas.push(...subSchema.type);
            } else {
                simpleSchemas.push((subSchema as Schema).type);
            }
            nullable = nullable === undefined ? subSchema.nullable : nullable || subSchema.nullable;
        } else if (isSimpleSchemaWithNullable(subSchema) && options?.restrictiveNulls) {
            if (nullable === undefined) {
                if(Array.isArray(subSchema.type)) {
                    simpleSchemas.push(...subSchema.type);
                } else {
                    simpleSchemas.push((subSchema as Schema).type);
                }
                nullable = subSchema.nullable;
            } else if (nullable === subSchema.nullable) {
                if(Array.isArray(subSchema.type)) {
                    simpleSchemas.push(...subSchema.type);
                } else {
                    simpleSchemas.push((subSchema as Schema).type);
                }
            } else {
                complexSchemas.push(subSchema);
            }
        } else {
            complexSchemas.push(subSchema);
        }
    }
    if (complexSchemas.length === 0) {
        return { type: simpleSchemas, ...(nullable === undefined ? {} : (nullable ? {} : {nullable})) };
    }
    const anyOf = [];
    if (simpleSchemas.length > 0) {
        anyOf.push({ type: simpleSchemas.length > 1 ? simpleSchemas : simpleSchemas[0], ...(nullable === undefined ? {} : (nullable ? {} : {nullable}))});
    }
    anyOf.push(...complexSchemas);
    return { anyOf };
}

function intersection(valuesArr: string[][]) {
    if (valuesArr.length === 0) return [];
    const arrays = valuesArr.filter(Array.isArray);
    const counter: Record<string, number> = {};
    for (const arr of arrays) {
        for (const val of arr) {
            if (!counter[val]) {
                counter[val] = 1;
            } else {
                counter[val]++;
            }
        }
    }
    return Object.entries(counter)
        .filter(([_, value]) => value === arrays.length)
        .map(([key]) => key);
}

function isSimpleSchema(schema: Schema): boolean {
    const keys = Object.keys(schema);
    return keys.length === 1 && keys[0] === 'type';
}

function isSimpleSchemaWithNullable(schema: Schema): boolean {
    const keys = Object.keys(schema);
    return keys.length === 2 && keys.indexOf('type') > -1 && keys.indexOf('nullable') > -1;
}

function isContainerSchema(schema: Schema): boolean {
    const type = (schema as Schema).type;
    return type === ValueType.Array || type === ValueType.Object;
}

// FACADE

export function createSchema(value: any, options?: SchemaGenOptions): Schema {
    if (typeof value === 'undefined') value = null;
    const clone = JSON.parse(JSON.stringify(value));
    return createSchemaFor(clone, options);
}

export function mergeSchemas(schemas: Schema[], options?: SchemaGenOptions): Schema {
    return combineSchemas(schemas, options);
}

export function extendSchema(schema: Schema, value: any, options?: SchemaGenOptions): Schema {
    const valueSchema = createSchema(value, options);
    return combineSchemas([schema, valueSchema], options);
}

export function createCompoundSchema(values: any[], options?: SchemaGenOptions): Schema {
    const schemas = values.map((value) => createSchema(value, options));
    return mergeSchemas(schemas, options);
}
