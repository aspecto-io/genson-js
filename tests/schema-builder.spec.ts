import { createSchema, mergeSchemas, ValueType, extendSchema, createCompoundSchema } from '../src';
import { pp } from './test-utils';

describe('SchemaBuilder', () => {
    describe('generation', () => {
        describe('simple types', () => {
            it('should build schema for integer', () => {
                const schema = createSchema(1);
                expect(schema).toEqual({ type: 'integer' });
            });

            it('should build schema for number', () => {
                const schema = createSchema(1.1);
                expect(schema).toEqual({ type: 'number' });
            });

            it('should build schema for string', () => {
                const schema = createSchema('some string');
                expect(schema).toEqual({ type: 'string' });
            });

            it('should build schema for null', () => {
                const schema = createSchema(null);
                expect(schema).toEqual({ type: 'null' });
            });

            it('should build schema for boolean', () => {
                const schema = createSchema(false);
                expect(schema).toEqual({ type: 'boolean' });
            });

            it('should build schema for array', () => {
                const schema = createSchema([]);
                expect(schema).toEqual({ type: 'array' });
            });

            it('should build schema for object', () => {
                const schema = createSchema({});
                expect(schema).toEqual({ type: 'object' });
            });

            it('should build schema for undefined', () => {
                const schema = createSchema(undefined);
                expect(schema).toEqual({ type: 'null' });
            });
        });

        describe('arrays', () => {
            it('it should generate schema for arrays of the same type', () => {
                const schema = createSchema([1, 2, 3]);
                expect(schema).toEqual({ type: 'array', items: { type: 'integer' } });
            });

            it('it should generate schema for arrays of the same type with floats', () => {
                const schema = createSchema([1, 2.1, 3]);
                expect(schema).toEqual({ type: 'array', items: { type: 'number' } });
            });

            it('it should generate schema for arrays of different primitive types', () => {
                const schema = createSchema([1, 1.1, 'string', null, false, true]);
                expect(schema).toEqual({ type: 'array', items: { type: ['null', 'boolean', 'number', 'string'] } });
            });

            it('it should generate schema for arrays of different primitive types and ints only', () => {
                const schema = createSchema([1, 'string', null, false, true]);
                expect(schema).toEqual({ type: 'array', items: { type: ['null', 'boolean', 'integer', 'string'] } });
            });
        });

        describe('objects', () => {
            it('it should generate schema for object with props of the same type', () => {
                const schema = createSchema({ one: 1, two: 2 });
                expect(schema).toEqual({
                    type: 'object',
                    properties: { one: { type: 'integer' }, two: { type: 'integer' } },
                    required: ['one', 'two'],
                });
            });

            it('it should generate schema for object with props of different types', () => {
                const schema = createSchema({ one: 1, two: 'second' });
                expect(schema).toEqual({
                    type: 'object',
                    properties: { one: { type: 'integer' }, two: { type: 'string' } },
                    required: ['one', 'two'],
                });
            });

            it('it should generate schema for object with props of different types w/o required', () => {
                const schema = createSchema({ one: 1, two: 'second' }, { noRequired: true });
                expect(schema).toEqual({
                    type: 'object',
                    properties: { one: { type: 'integer' }, two: { type: 'string' } },
                });
            });
            it('it should generate schema for object with additionalProperties=false ', () => {
                const schema = createSchema({ one: 1, two: 'second' }, { noAdditionalProperties: true });
                expect(schema).toEqual({
                    type: 'object',
                    additionalProperties: false,
                    properties: { one: { type: 'integer' }, two: { type: 'string' } },
                    required: [
                        "one",
                        "two",
                    ]
                });
            });
        });

        describe('nested array', () => {
            it('should generate schema for nested arrays', () => {
                const schema = createSchema([1, [2.1], [[3]]]);
                expect(schema).toEqual({
                    type: 'array',
                    items: {
                        anyOf: [
                            {
                                type: 'integer',
                            },
                            {
                                type: 'array',
                                items: {
                                    anyOf: [
                                        {
                                            type: 'number',
                                        },
                                        {
                                            type: 'array',
                                            items: {
                                                type: 'integer',
                                            },
                                        },
                                    ],
                                },
                            },
                        ],
                    },
                });
            });

            it('should generate schema for nested arrays and simplify anyOf', () => {
                const schema = createSchema([1, 'some string', null, [2, 'some other string', {}], [[3.1]]]);
                expect(schema).toEqual({
                    type: 'array',
                    items: {
                        anyOf: [
                            {
                                type: ['null', 'integer', 'string'],
                            },
                            {
                                type: 'array',
                                items: {
                                    anyOf: [
                                        {
                                            type: ['integer', 'string', 'object'],
                                        },
                                        {
                                            type: 'array',
                                            items: {
                                                type: 'number',
                                            },
                                        },
                                    ],
                                },
                            },
                        ],
                    },
                });
            });
        });

        describe('nested objects/arrays', () => {
            it('should combine object schemas and respect required property', () => {
                const schema = createSchema({
                    one: 1,
                    two: 'second',
                    three: { four: 5, five: [5], six: null, seven: [{}, { eight: 1.1 }, { nine: 'nine' }] },
                });
                expect(schema).toMatchSnapshot();
            });

            it('should generate schema for nested object with props of different types w/o required', () => {
                const schema = createSchema({ one: 1, two: { a: 'value' } }, { noRequired: true });
                expect(schema).toEqual({
                    type: 'object',
                    properties: {
                        one: { type: 'integer' },
                        two: { type: 'object', properties: { a: { type: 'string' } } },
                    },
                });
            });
        });

        describe('all cases combined', () => {
            it('should generate valid schemas for complex objects', () => {
                const schema = createSchema([
                    {
                        lvl1PropNum: 1,
                        lvl1PropStr: 'second',
                        lvl1PropObj1: { lvl2PropArr: [1, 2] },
                        lvl1PropObj2: {
                            lvl2PropNum1: 5,
                            lvl2PropArr1: [5],
                            six: null,
                            lvl2PropArr2: [
                                {},
                                { lvl3PropNum1: 1.2 },
                                { lvl3PropStr1: 'nine' },
                                [1, false],
                                1,
                                'some string',
                            ],
                        },
                    },
                    { lvl1PropStr: 'one' },
                    { lvl1PropNum: 1.2, lvl1PropStr: 'one' },
                    { lvl1PropStr: 'one', lvl1PropObj1: { lvl2PropArr: [2.3, null, 'some string', false] } },
                ]);

                expect(schema).toMatchSnapshot();
            });

            it('should consider value as required if it is present in all objects', async () => {
                const val = [
                    {
                        arr: [
                            {
                                prop1: 'test string',
                            },
                            {
                                prop2: 'test string',
                            },
                        ],
                    },
                    {
                        arr: [
                            {
                                prop1: 'test',
                            },
                        ],
                    },
                ];
                const schema = createSchema(val);
                expect(schema).toEqual({
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            arr: {
                                type: 'array',
                                items: {
                                    type: 'object',
                                    properties: {
                                        prop1: {
                                            type: 'string',
                                        },
                                        prop2: {
                                            type: 'string',
                                        },
                                    },
                                },
                            },
                        },
                        required: ['arr'],
                    },
                });
            });

            it('should consider value with additionalProperties=false in all objects', async () => {
                const val = [
                    {
                        arr: [
                            {
                                prop1: 'test string',
                            },
                            {
                                prop2: 'test string',
                            },
                        ],
                    },
                    {
                        arr: [
                            {
                                prop1: 'test',
                            },
                        ],
                    },
                ];
                const schema = createSchema(val, { noAdditionalProperties: true });
                expect(schema).toEqual({
                    type: 'array',
                    items: {
                        type: 'object',
                        additionalProperties: false,
                        properties: {
                            arr: {
                                type: 'array',
                                items: {
                                    type: 'object',
                                    additionalProperties: false,
                                    properties: {
                                        prop1: {
                                            type: 'string',
                                        },
                                        prop2: {
                                            type: 'string',
                                        },
                                    },
                                },
                            },
                        },
                        required: ['arr'],
                    },
                });
            });

            it('should generate schema for array of objects w/o required', () => {
                const schema = createSchema(
                    [
                        { one: 'a', two: 'b' },
                        { one: 'aa', two: 'bb' },
                    ],
                    { noRequired: true }
                );
                expect(schema).toEqual({
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            one: { type: 'string' },
                            two: { type: 'string' },
                        },
                    },
                });
            });
        });

        describe('prototype methods', () => {
            function createObjectWithProtoMethods(): Record<string, any> {
                return {
                    constructor: 'constructor value',
                    hasOwnProperty: 'hasOwnProperty value',
                    isPrototypeOf: 'isPrototypeOf value',
                    propertyIsEnumerable: 'propertyIsEnumerable value',
                    toLocaleString: 'toLocaleString value',
                    toString: 'toString value',
                    valueOf: 'valueOf value',
                    __defineGetter__: '__defineGetter__ value',
                    __defineSetter__: '__defineSetter__ value',
                    __lookupGetter__: '__lookupGetter__ value',
                    __lookupSetter__: '__lookupSetter__ value',
                    __proto__: '__proto__ value',
                };
            }

            it('should work for props with the same names as Object.prototype methods', async () => {
                // it's improtant to keep them in array, as this is more complex case
                const value: any = [createObjectWithProtoMethods(), createObjectWithProtoMethods()];
                const schema = createSchema(value);
                expect(schema).toEqual({
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            constructor: { type: 'string' },
                            hasOwnProperty: { type: 'string' },
                            isPrototypeOf: { type: 'string' },
                            propertyIsEnumerable: { type: 'string' },
                            toLocaleString: { type: 'string' },
                            toString: { type: 'string' },
                            valueOf: { type: 'string' },
                            __defineGetter__: { type: 'string' },
                            __defineSetter__: { type: 'string' },
                            __lookupGetter__: { type: 'string' },
                            __lookupSetter__: { type: 'string' },
                        },
                    },
                });
            });
        });

        describe('circular refs', () => {
            it('should throw an error w/ explanation', async () => {
                const a: any = {};
                const b: any = {};
                a.b = b;
                b.a = a;

                expect(() => {
                    createSchema(a);
                }).toThrowErrorMatchingSnapshot();
            });
        });

        describe('non-json values', () => {
            it('should ignore functions', async () => {
                const value: any = {
                    func() {},
                    someProp: 'string',
                };
                const schema = createSchema(value);
                expect(schema).toEqual({
                    type: 'object',
                    properties: {
                        someProp: {
                            type: 'string',
                        },
                    },
                    required: ['someProp'],
                });
            });
        });
    });

    describe('createCompoundSchema', () => {
        it('should create compound schema from multiple inputs', () => {
            const schema = createCompoundSchema([{ age: 35 }, { age: 19, name: 'John' }, { age: 23, admin: true }]);
            expect(schema).toEqual({
                type: 'object',
                properties: { admin: { type: 'boolean' }, age: { type: 'integer' }, name: { type: 'string' } },
                required: ['age'],
            });
        });
    });

    describe('merging', () => {
        it('should merge simple schemas', () => {
            const merged = mergeSchemas([{ type: ValueType.Number }, { type: ValueType.String }]);
            expect(merged).toEqual({ type: ['number', 'string'] });
        });

        it('should merge array schemas', () => {
            const merged = mergeSchemas([
                {
                    type: ValueType.Array,
                    items: {
                        anyOf: [
                            {
                                type: [ValueType.Number, ValueType.String],
                            },
                            {
                                type: ValueType.Array,
                            },
                        ],
                    },
                },
                { type: ValueType.String },
            ]);
            expect(merged).toEqual({
                anyOf: [
                    { type: ValueType.String },
                    {
                        type: ValueType.Array,
                        items: { type: [ValueType.Number, ValueType.String, ValueType.Array] },
                    },
                ],
            });
        });

        it('should merge object schemas', async () => {
            const merged = mergeSchemas([
                { type: ValueType.Object },
                { type: ValueType.Object, properties: { prop1: { type: ValueType.String } } },
            ]);
            expect(merged).toEqual({
                type: ValueType.Object,
                properties: { prop1: { type: ValueType.String } },
            });
        });

        it('should merge more than 2 schemas', async () => {
            const merged = mergeSchemas([
                { type: ValueType.Number },
                { type: ValueType.String },
                { type: ValueType.Boolean },
            ]);
            expect(merged).toEqual({ type: ['boolean', 'number', 'string'] });
        });
    });

    describe('extending', () => {
        it('should extend simple schemas', () => {
            const merged = extendSchema({ type: ValueType.Number }, 'some string');
            expect(merged).toEqual({ type: ['number', 'string'] });
        });

        it('should extend array schemas', () => {
            const merged = extendSchema(
                {
                    type: ValueType.Array,
                    items: {
                        anyOf: [
                            {
                                type: [ValueType.Number, ValueType.String],
                            },
                            {
                                type: ValueType.Array,
                            },
                        ],
                    },
                },
                'some string'
            );
            expect(merged).toEqual({
                anyOf: [
                    { type: ValueType.String },
                    {
                        type: ValueType.Array,
                        items: { type: [ValueType.Number, ValueType.String, ValueType.Array] },
                    },
                ],
            });
        });
    });
});
