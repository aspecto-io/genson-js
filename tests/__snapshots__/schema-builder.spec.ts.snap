// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaBuilder generation all cases combined should generate valid schemas for complex objects 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "lvl1PropNum": Object {
        "type": "number",
      },
      "lvl1PropObj1": Object {
        "properties": Object {
          "lvl2PropArr": Object {
            "items": Object {
              "type": Array [
                "null",
                "boolean",
                "number",
                "string",
              ],
            },
            "type": "array",
          },
        },
        "required": Array [
          "lvl2PropArr",
        ],
        "type": "object",
      },
      "lvl1PropObj2": Object {
        "properties": Object {
          "lvl2PropArr1": Object {
            "description": "some description",
            "items": Object {
              "type": "integer",
            },
            "type": "array",
          },
          "lvl2PropArr2": Object {
            "items": Object {
              "anyOf": Array [
                Object {
                  "type": Array [
                    "integer",
                    "string",
                  ],
                },
                Object {
                  "items": Object {
                    "type": Array [
                      "boolean",
                      "integer",
                    ],
                  },
                  "type": "array",
                },
                Object {
                  "properties": Object {
                    "lvl3PropNum1": Object {
                      "type": "number",
                    },
                    "lvl3PropStr1": Object {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
          "lvl2PropNum1": Object {
            "type": "integer",
          },
          "six": Object {
            "type": "null",
          },
        },
        "required": Array [
          "lvl2PropNum1",
          "lvl2PropArr1",
          "six",
          "lvl2PropArr2",
        ],
        "type": "object",
      },
      "lvl1PropStr": Object {
        "type": "string",
      },
    },
    "required": Array [
      "lvl1PropStr",
    ],
    "type": "object",
  },
  "type": "array",
}
`;

exports[`SchemaBuilder generation circular refs should throw an error w/ explanation 1`] = `
"Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    |     property 'b' -> object with constructor 'Object'
    --- property 'a' closes the circle"
`;

exports[`SchemaBuilder generation nested objects/arrays should combine object schemas and respect required property 1`] = `
Object {
  "properties": Object {
    "one": Object {
      "type": "integer",
    },
    "three": Object {
      "properties": Object {
        "five": Object {
          "items": Object {
            "type": "integer",
          },
          "type": "array",
        },
        "four": Object {
          "type": "integer",
        },
        "seven": Object {
          "items": Object {
            "properties": Object {
              "eight": Object {
                "type": "number",
              },
              "nine": Object {
                "type": "string",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
        "six": Object {
          "type": "null",
        },
      },
      "required": Array [
        "four",
        "five",
        "six",
        "seven",
      ],
      "type": "object",
    },
    "two": Object {
      "type": "string",
    },
  },
  "required": Array [
    "one",
    "two",
    "three",
  ],
  "type": "object",
}
`;
